### ISSUE TITLE
refactor: further replace handwritten RL infrastructure with TorchRL primitives

### ISSUE THREAD

**Issue by xcz0** [xcz0/Eco-AutoDrive#70]
## 背景

在 #68 的 evaluation/runtime 重构基础上，当前 RL 已经采用：

- `TensorDict`
- `torchrl.objectives.value.GAE`
- `torchrl.objectives.ClipPPOLoss`
- `ProbabilisticTensorDictModule` / `TensorDictModule`

因此 PPO 核心数学已经不再手写。下一步更值得替换的是 PPO 外围基础设施：minibatch/storage、rollout transition contract、policy TensorDict glue、collector/vector env。

本 Issue 的目标不是“为了使用 TorchRL 而使用 TorchRL”，而是在保持现有研究语义、RNG 可复现性和 artifact/checkpoint contract 的前提下，进一步删除可由成熟 TorchRL primitive 承担的通用基础设施。

关联：#68

### 1. PPO minibatch / storage

`src/eco_planner/rl/ppo.py` 仍手动实现：

- `torch.randperm()` epoch shuffle；
- `_epoch_minibatch_indices()`；
- minibatch slicing；
- minibatch RNG state 持久化。

这部分可以评估迁移到：

```text
TensorDictReplayBuffer
LazyTensorStorage
SamplerWithoutReplacement(drop_last=True)
```

需要保持当前 contract：每个 PPO epoch 中，每个 full-batch sample 恰好被消费一次，不跨 epoch 尾部补样本。

### 2. rollout transition / bootstrap boundary

`src/eco_planner/rl/rollout.py` 当前手动：

- 累积 `training_transitions`；
- `torch.cat()` episode；
- 构造 `next` TensorDict；
- 左移 `state_value`；
- 注入 tail bootstrap value；
- 手工维护 `done / terminated / truncated` 与 `tail_kind`。

可以进一步靠近 TorchRL 标准 transition layout：

```text
root:
    observation / policy context
    action
    state_value
next:
    observation
    reward
    done
    terminated
    truncated
```

但必须保持 ADR 0018 的 episode boundary：

- true `terminated` 不 bootstrap；
- simulator `truncated` 使用显式 tail value bootstrap；
- `rollout_limit` 使用显式 tail value bootstrap；
- GAE 不得跨 episode boundary。

第一阶段仍建议 per-episode 执行 GAE，不将不同 episode 直接 flatten 后统一递归。

### 3. Exploration Policy 的 TensorDict glue

当前 policy 主接口仍是：

```text
ExplorationPolicyContext dataclass
 -> ExplorationPolicy
 -> ExplorationPolicyOutput dataclass
```

而 PPO 侧额外维护：

- `_PolicyOutputAdapter`
- `_CONTEXT_KEYS`
- `TensorDictModule`
- `ProbabilisticTensorDictModule`
- identity critic adapter

trainer/runtime 中也存在多处 dataclass ↔ TensorDict/Tensor 的序列化、device transfer 和 `unsqueeze()` glue。

可考虑把 shared trunk / actor head / value head 提升为 TensorDict-native module boundary，使同一 policy contract 同时服务：

- rollout inference；
- PPO current-policy log-prob recomputation；
- value evaluation；
- checkpoint probe。

保留 Exploration Policy 本身的网络结构和参数共享语义。

### 4. EnvBase / TensorSpec

当前 rollout/runtime 中存在大量结构性验证：

- tensor shape；
- dtype；
- device；
- observation batch size；
- action/reward/done shape。

可以增加一个薄的 TorchRL `EnvBase` adapter，并定义 observation/action/reward/done specs，使用 `check_env_specs()` 覆盖通用 tensor contract。

但以下项目领域约束继续保留手工检查：

- guidance action 必须严格位于 `(-1, 1)`；
- reward audit 等式；
- RL rollout 每 transition 必须执行恰好 1 个 0.1 s simulator substep；
- evaluation 每 cycle 仍执行前 5 个 trajectory points；
- traffic warmup / route / artifact 等研究语义。

原则：TorchRL spec 只替代“数据结构合法性”，不替代“研究语义合法性”。

### 5. Collector

`VectorRolloutCollector` 当前手动维护：

- per-slot observations；
- training/audit transition lists；
- transition counts；
- episode cycles；
- tail detection；
- bootstrap slot batching；
- episode finalize/reset；
- physical-slot wave scheduling。

在 EnvBase/TensorDict transition contract 稳定后，可以评估 TorchRL `Collector` / synchronous multi-collector 是否能接管标准 policy→step→reset→stack orchestration。

PPO 必须保持 on-policy，同步 collection 优先，不允许由于异步 collector 引入 policy lag。

### 6. `VectorMetaDriveEnv` process/IPC

`src/eco_planner/envs/vector_metadrive.py` 当前自行实现：

- `multiprocessing.Process`；
- `Pipe` / `Connection`；
- send/receive protocol；
- partial step/reset；
- worker failure propagation；
- join/terminate 生命周期。

TorchRL `ParallelEnv` 在概念上可以覆盖固定进程环境、partial step/reset 等基础设施。

但该 worker 还承担：

- traffic history；
- warmup；
- observation adapter；
- map replacement/cache；
- route metadata；
- trajectory execution audit。

因此本项只做 PoC，不作为前置迁移目标。只有 exact behavior、failure semantics 和 throughput 均满足要求后再考虑删除自定义 IPC wrapper。

### Explicit-generator affine Beta sampling

不要直接用标准 TorchRL stochastic actor 替代 collection 阶段的 `AffineBeta` / `ExplicitGeneratorBetaSampler`。

当前实现要求：

- 每个 rollout slot 有独立 `torch.Generator`；
- policy RNG state 可保存并重放；
- affine Beta action 在 transformed guidance space 中计算正确 joint log-prob / entropy。

TorchRL probabilistic module继续用于 PPO current-policy log-prob recomputation即可；collection 仍保留 explicit-generator sampling。

### Advantage normalization

暂不直接启用：

```python
ClipPPOLoss(normalize_advantage=True)
```

现有 ADR 0018 contract 要求：

- 整个当前 PPO batch 统一标准化一次；
- 使用 sample standard deviation；
- zero variance / non-finite 直接失败；
- 不通过 clamp 隐藏退化 batch。

除非验证 TorchRL 内置 normalization 与该 contract 完全等价，否则继续保留显式 full-batch normalization。

### Diffusion / MetaDrive / evaluation artifact semantics

以下不纳入 TorchRL 替换：

- Diffusion Planner inference；
- DDIM / guidance RNG；
- MetaDrive trajectory execution；
- evaluation 5 × 0.1 s execution semantics；
- trace / summary / artifact schema；
- failure persistence；
- project-specific audit fields。

### Phase 1 — ReplayBuffer / sampler

- [ ] 使用 `TensorDictReplayBuffer` + `LazyTensorStorage` + `SamplerWithoutReplacement(drop_last=True)` 替换 `_epoch_minibatch_indices()`。
- [ ] 删除手工 permutation/chunk logic。
- [ ] 明确并测试 sampler RNG 的 resumable checkpoint contract。
- [ ] 增加 fixed-seed minibatch sequence regression test。

### Phase 2 — TensorDict-native transition

- [ ] rollout transition 从产生时即采用标准 root/`next` layout。
- [ ] 减少 `finalize_rollout_episode()` 中手工 value shift / next 拼装。
- [ ] 保留 `terminated / truncated / rollout_limit` bootstrap distinction。
- [ ] 保留 per-episode GAE boundary。

### Phase 3 — TensorDict-native policy boundary

- [ ] 减少 `_PolicyOutputAdapter` 和重复 context glue。
- [ ] shared trunk / actor/value outputs 通过统一 TensorDict keys 暴露。
- [ ] rollout、PPO、bootstrap value 和 probe 共用同一 policy tensor contract。
- [ ] explicit-generator Beta collection sampling 保持不变。

### Phase 4 — EnvBase + specs

- [ ] 为 MetaDrive RL boundary 增加薄 `EnvBase` adapter。
- [ ] 定义 observation/action/reward/done specs。
- [ ] 使用 `check_env_specs()` 替代重复结构性验证。
- [ ] 项目领域 invariant 继续显式验证。

### Phase 5 — Collector PoC

- [ ] 先以 no-traffic smoke 验证 TorchRL collector。
- [ ] 保证 PPO synchronous on-policy semantics。
- [ ] 保证 per-slot diffusion/policy RNG stream 不变化。
- [ ] 保证 audit/replay payload 不丢失。
- [ ] 与当前 collector 做 exact rollout comparison。

### Phase 6 — ParallelEnv PoC

- [ ] 验证 partial step/reset。
- [ ] 验证 map replacement、warmup、traffic history 和 execution audit。
- [ ] 验证 worker crash/failure propagation。
- [ ] 比较 throughput / IPC overhead。
- [ ] 仅在无行为退化时考虑删除 `VectorMetaDriveEnv` 自定义 multiprocessing 基础设施。

## 验收标准

- [ ] PPO 数值语义与 ADR 0018 一致。
- [ ] rollout reward / termination / bootstrap 语义不变。
- [ ] diffusion RNG 与 policy RNG replay contract 不变。
- [ ] frozen planner 不被 PPO 更新。
- [ ] checkpoint resume 后 future minibatch / rollout RNG 可复现。
- [ ] 现有 rollout artifact / smoke tests 通过。
- [ ] evaluation numerical/artifact semantics 不受本 Issue 影响。
- [ ] 删除的代码主要是通用 RL infrastructure，而不是把项目特有逻辑隐藏进 TorchRL wrapper。

## 优先级

建议优先级：

```text
P0  ReplayBuffer / Sampler
P0  standard TensorDict transition contract
P1  TensorDict-native policy modules
P1  EnvBase / specs
P2  Collector
P2  ParallelEnv
```

核心原则：先替换边界清晰、风险低的成熟 primitive，再逐步评估 collector/vector env；不要一次性重写整个 rollout stack。
